Objects/regions.c

```diff
@@ -92,9 +116,324 @@ struct regionmetadata {
+        assert(!IS_CONST_REGION(other) && "Constants can never have children");
+
+        regionmetadata_set_parent(self, NULL);
+
+        goto merge;
+    }
+
+    // If only `self` has a parent we can make `other` the child and
+    // remove the parent from `self`. The merged region will then again
+    // ahev the correct parent.
+    regionmetadata* other_parent = regionmetadata_get_parent(self);
+    if (!IS_CONST_REGION(other) && IS_REGION_NULL(other_parent)) {
+        regionmetadata_set_parent(_Py_CAST(regionmetadata*, other), self_parent);
+        regionmetadata_set_parent(self, NULL);
+
+        goto merge;
+    }
+
+    PyObject* other_node = NULL;
+    if (!IS_REGION_NULL(other) && !IS_CONST_REGION(other))  {
+        other_node = _PyObject_CAST(_Py_CAST(regionmetadata*, other)->bridge);
+    }
+    throw_region_error(self->bridge, other_node,
+                       "Unable to merge regions", NULL);
+    return NULL;
+
+merge:
+    regionmetadata_inc_rc(other);
+
+    // Move LRC and OSC into the root.
+    if (!IS_REGION_NULL(other) && !IS_CONST_REGION(other)) {
+        // Move information into the merge root
+        regionmetadata* other_data = _Py_CAST(regionmetadata*, other);
+        other_data->lrc += self->lrc;
+        other_data->osc += self->osc;
+        other_data->is_open |= self->is_open;
+        // remove information from self
+        self->lrc = 0;
+        self->osc = 0;
+        self->is_open = false;
+    }
+
+    self->parent = Py_region_ptr_with_tags(other | Py_METADATA_MERGE_TAG);
+    // No decref, since this is a week reference. Otherwise we would get
+    // a cycle on a non GC'ed object and the bridge.
+    self->bridge = NULL;
+    Py_RETURN_NONE;
+}
+#define regionmetadata_merge(self, other) \
+  (regionmetadata_merge(self, _Py_CAST(Py_region_ptr_t, other)));
+
+static bool is_bridge_object(PyObject *op) {
+    regionmetadata* region = regionmetadata_get_root_or_null(Py_REGION(op));
+    // Constant regions (indicated by NULL) never have a bridge object.
+    if (!region) {
+        return false;
+    }
+
+    // It's not yet clear how immutability will interact with region objects.
+    // It's likely that the object will remain in the object topology but
+    // will use the properties of a bridge object. This therefore checks if
+    // the object is equal to the regions bridge object rather than checking
+    // that the type is `PyRegionObject`
+    return _PyObject_CAST(region->bridge) == op;
+}
+
+static bool regionmetadata_has_ancestor(regionmetadata* self, regionmetadata* other) {
+    other = regionmetadata_get_root_or_null(other);
+
+    // Constant regions can't be an ancestor
+    if (!other) {
+        return false;
+    }
+
+    // Make sure we actually look at our representative
+    self = regionmetadata_get_root_or_null(self);
+    while (self) {
+        if (self == other) {
+            return true;
+        }
+        self = regionmetadata_get_parent(self);
```

Comment:
Is this correct? Don't we need to consider the representative also for all parent regions?